### PR TITLE
Add logic for allreduce failure handling in Worker

### PR DESCRIPTION
**Pull request by terrytangyuan**
Based on design doc proposed in #1420.

### CODE REVIEW

File: elasticdl/python/worker/worker.py [hunk 1]
```diff
@@ -546,17 +555,62 @@ def forward_process(self, features):
...
+
+    def _get_local_model_params(self):
+        return {}
+
+    def _broadcast_model_params(self):
+        status = self._collective_communicator.barrier()
+        if status == CollectiveCommunicatorStatus.FAILED:
+            logger.warning("Failed to perform barrier operation")
+            return False
+        root_ip = random.choice(
```
Comment (skydoorkai) [inline elasticdl/python/worker/worker.py]: If each worker selects different `root_ip`, `broadcast` will not work.

File: elasticdl/python/worker/worker.py [hunk 2]
```diff
@@ -569,7 +623,18 @@ def _collect_gradients_without_allreduce(self, grads):
     def _run_training_task(self, features, labels):
         loss, grads = self.training_process(features, labels)
         if self._distribution_strategy == DistributionStrategy.ALLREDUCE:
-            return (*self._collect_gradients_with_allreduce(grads), loss)
+            # TODO: Delay certain amount of time before retrying
+            for _ in range(self._max_allreduce_retry_num + 1):
+                accepted, loss = self._collect_gradients_with_allreduce_robust(
```
Comment (skydoorkai) [inline elasticdl/python/worker/worker.py]: Why get `loss` here?

File: elasticdl/python/worker/worker.py (hunk 1 above)
Comment (skydoorkai) [reply elasticdl/python/worker/worker.py]: Need to choose an old worker?

File: elasticdl/python/worker/worker.py (hunk 2 above)
Comment (terrytangyuan) [reply elasticdl/python/worker/worker.py]: Thanks. Fixed!

File: elasticdl/python/worker/worker.py (hunk 1 above)
Comment (terrytangyuan) [reply elasticdl/python/worker/worker.py]: Good catch! This worker should be selected by master. I just moved this from `CollectiveCommunicator` to `Worker._get_broadcast_root_worker_ip()` and added TODO notes.

**Review by skydoorkai — APPROVED** [review]
